Automattic/wp-calypso
# Media: Add state behaviors for series upload

~~Blocks #9675~~
Related: #5046

This pull request seeks to start implementation of media Redux state handlers. The included behaviors are quite basic, including only uploads of one or many files, without validation or compatibility with prior media Flux state.

__Implementation Notes:__

Recognizing a common pattern of keying reducer state objects by site ID, included in the changes is a proposed `createSiteReducer`, an enhanced variant of `createReducer` which automatically keys the state returned from the reducer by an assumed `siteId` property on the action.

__Testing Instructions:__

Ensure that Mocha tests pass:

```
npm run test-client
```

The commit 6922b25 is a temporary and unfinished demonstration of expected behaviors. It replaces existing Flux dispatching with the new Redux action creators usage. Verify that after uploading one or many files to the editor media library that the uploads occur in series order (i.e. uploads one at a time). This is easiest to verify using the [Redux DevTools extension](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=en), but you can also use the newly introduced [action logging feature](https://github.com/Automattic/wp-calypso/tree/master/client/state/console-dispatch).

![upload](https://cloud.githubusercontent.com/assets/1779930/20729793/96250e14-b651-11e6-85bc-060c28dbc572.gif)

1. Navigate to the [post editor](http://calypso.localhost:3000/post)
2. Select a site, if prompted
3. Click Add Media in the editor toolbar
4. Upload one or more media items
5. Observe that Redux actions are logged in expected sequence
6. Refresh the page (items will **not** show in library in same session yet)
7. Click Add Media in the editor toolbar
8. Note that items were uploaded successfully

cc @tyxla @gwwar
